train.py

```diff
@@ -0,0 +1,193 @@
+import os
+import uuid
+
+import hydra
+import torch
+import wandb
+import torch.nn as nn
+import torch.optim as optim
+import pytorch_lightning as pl
+from datasets import load_dataset
+from torch.utils.data import DataLoader
+from pytorch_lightning.callbacks import TQDMProgressBar
+from pytorch_lightning.loggers.wandb import WandbLogger
+
+from config import Config
+from models.utils import Transforms
+from models.reverse_diffusion import Unet
+from models.forward_diffusion import ForwardDiffusion
+
+
+class UnetTrainingWrapper(pl.LightningModule):
+    def __init__(
+        self,
+        model: Unet,
+        forward_diffusion: ForwardDiffusion,
+        lr: float = 1e-3,
+        l2: float = 0.01,
+        loss_type: str = "mse",
+        ckpt_dir: str = "checkpoints",
+        save_every_n_steps: int = 100,
+    ):
+        super().__init__()
+
+        # model
+        self.model = model
+        self.forward_diffusion = forward_diffusion
+        self.timesteps = forward_diffusion.timesteps
+
+        # save hyperparameters (lr, l2, loss_type, ckpt_dir, save_every_n_steps)
+        self.save_hyperparameters(ignore=["model", "forward_diffusion"])
+
+        # get loss function
+        self.loss_function_module = self._loss_function()
+
+    def _loss_function(self) -> nn.Module:
+        if self.hparams["loss_type"] == "mse":
+            loss = nn.MSELoss()
+        elif self.hparams["loss_type"] == "mae":
+            loss = nn.L1Loss()
+        elif self.hparams["loss_type"] == "huber":
+            loss = nn.HuberLoss()
+        else:
+            raise NotImplementedError()
+
+        return loss
+
+    def _save_models(self):
+        # ckpt specifies directory and name of the file is name of the experiment in wandb
+        save_path = f"{self.hparams['ckpt_dir']}/{self.logger.experiment.name}.ckpt"
+        # saving models
+        torch.save({"model": self.model.state_dict(), "forward_diffusion": self.forward_diffusion.state_dict()}, save_path)
+
+    def forward(self, x: torch.Tensor, t: torch.Tensor) -> torch.Tensor:
+        return self.model(x, t)
+
+    def configure_optimizers(self) -> optim.Optimizer:
+        optimizer = optim.AdamW(self.model.parameters(), lr=self.hparams["lr"], weight_decay=self.hparams["l2"])
```

Comment:
```suggestion
        optimizer = optim.AdamW(
          params=self.model.parameters(),
          lr=self. hparams["lr"],
          weight_decay=self.hparams["l2"],
        )
```

My personal style-guide is to always prefer multi-line argument-per-line approach in cases like this. Switching from vertical to horizontal takes effort that we/I can save by breaking it out. Also, the diffs are much better, especially if we use the magic comma :)